File: ios/bookz/bookz/network/BooksModel.swift
//
//  BooksModel.swift
//  bookz
//
//  Created by Aldana, Sal on 1/15/23.

import Foundation

// MARK: - Single Book Response
struct BookResponse: Codable {
    let kind, id, etag: String?
    let selfLink: String?
    let volumeInfo: VolumeInfo?
    let saleInfo: SaleInfo?
    let accessInfo: AccessInfo?
    
    func convertToModel() -> BooksModel {
        let model = BooksModel(kind: self.kind,
                               id: self.id,
                               etag: self.etag,
                               selfLink: self.selfLink,
                               volumeInfo: self.volumeInfo,
                               saleInfo: self.saleInfo,
                               accessInfo: self.accessInfo,
                               searchInfo: nil)
        return model
    }
}

// MARK: - BooksResponse
struct BooksResponse: Codable {
    let kind: String
    let totalItems: Int
    let items: [BooksModel]
}

// MARK: - BooksModel
struct BooksModel: Codable, Hashable, Identifiable {
    let kind, id, etag: String?
    let selfLink: String?
    let volumeInfo: VolumeInfo?
    let saleInfo: SaleInfo?
    let accessInfo: AccessInfo?
    let searchInfo: SearchInfo?
    
    static func == (lhs: BooksModel, rhs: BooksModel) -> Bool {
        return lhs.id == rhs.id
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - AccessInfo
struct AccessInfo: Codable {
    let country, viewability: String?
    let embeddable, publicDomain: Bool?
    let textToSpeechPermission: String?
    let epub: Epub?
    let pdf: PDF?
    let webReaderLink: String?
    let accessViewStatus: String?
    let quoteSharingAllowed: Bool?
}

// MARK: - Epub
struct Epub: Codable {
    let isAvailable: Bool?
}

// MARK: - PDF
struct PDF: Codable {
    let isAvailable: Bool?
    let acsTokenLink: String?
}

// MARK: - SaleInfo
struct SaleInfo: Codable {
    let country, saleability: String?
    let isEbook: Bool?
}

// MARK: - SearchInfo
struct SearchInfo: Codable {
    let textSnippet: String?
}

// MARK: - VolumeInfo
struct VolumeInfo: Codable {
    let title: String?
    let authors: [String]?
    let publisher, publishedDate, description: String?
    let industryIdentifiers: [IndustryIdentifier]?
    let readingModes: ReadingModes?
    let pageCount: Int?
    let printType: String?
    let categories: [String]?
    let averageRating: Double?
    let ratingsCount: Int?
    let maturityRating: String?
    let allowAnonLogging: Bool?
    let contentVersion: String?
    let panelizationSummary: PanelizationSummary?
    let imageLinks: ImageLinks?
    let language: String?
    let previewLink, infoLink: String?
    let canonicalVolumeLink: String?
}

// MARK: - ImageLinks
struct ImageLinks: Codable {
    let smallThumbnail, thumbnail: String?
}

// MARK: - IndustryIdentifier
struct IndustryIdentifier: Codable {
    let type, identifier: String?
}

// MARK: - PanelizationSummary
struct PanelizationSummary: Codable {
    let containsEpubBubbles, containsImageBubbles: Bool?
}

// MARK: - ReadingModes
struct ReadingModes: Codable {
    let text, image: Bool?
}

File: android/app/src/main/java/net/level3studios/bookz/network/BooksModel.kt
package net.level3studios.bookz.network


data class BooksResponse(
    val kind: String,
    val totalItems: Int,
    val items: List<BooksModel>
)

data class BooksModel(
    val kind: String? = null,
    val id: String? = null,
    val etag: String? = null,
    val selfLink: String? = null,
    val volumeInfo: VolumeInfo? = null,
    val saleInfo: SaleInfo? = null,
    val accessInfo: AccessInfo? = null,
    val searchInfo: SearchInfo? = null
)

data class AccessInfo(
    val country: String? = null,
    val viewability: String? = null,
    val embeddable: Boolean? = null,
    val publicDomain: Boolean? = null,
    val textToSpeechPermission: String? = null,
    val epub: Epub? = null,
    val pdf: PDF? = null,
    val webReaderLink: String? = null,
    val accessViewStatus: String? = null,
    val quoteSharingAllowed: Boolean? = null
)

data class Epub(
    val isAvailable: Boolean? = null
)

data class PDF(
    val isAvailable: Boolean? = null,
    val acsTokenLink: String? = null
)

data class SaleInfo(
    val country: String? = null,
    val saleability: String? = null,
    val isEbook: Boolean? = null
)

data class SearchInfo(
    val textSnippet: String? = null
)

data class VolumeInfo(
    val title: String? = null,
    val authors: List<String>? = null,
    val publisher: String? = null,
    val publishedDate: String? = null,
    val description: String? = null,
    val industryIdentifiers: List<IndustryIdentifier>? = null,
    val readingModes: ReadingModes? = null,
    val pageCount: Long? = null,
    val printType: String? = null,
    val categories: List<String>? = null,
    val averageRating: Double? = null,
    val ratingsCount: Long? = null,
    val maturityRating: String? = null,
    val allowAnonLogging: Boolean? = null,
    val contentVersion: String? = null,
    val panelizationSummary: PanelizationSummary? = null,
    val imageLinks: ImageLinks? = null,
    val language: String? = null,
    val previewLink: String? = null,
    val infoLink: String? = null,
    val canonicalVolumeLink: String? = null
) {
    val ratingsCountString: String
        get() {
            return if (this.ratingsCount != null) {
                this.ratingsCount.toString()
            } else {
                "0"
            }
        }

    val averageRatingString: String
        get() {
            return if (this.averageRating != null) {
                this.averageRating.toString()
            } else {
                return "0.0"
            }
        }
}

data class ImageLinks(
    val smallThumbnail: String? = null,
    val thumbnail: String? = null
)

data class IndustryIdentifier(
    val type: String? = null,
    val identifier: String? = null
)

data class PanelizationSummary(
    val containsEpubBubbles: Boolean? = null,
    val containsImageBubbles: Boolean? = null
)

data class ReadingModes(
    val text: Boolean? = null,
    val image: Boolean? = null
)